File: study/ts/script/scene/view/grainCar/GrainCarConfigHelper.ts
import { ConfigNameConst } from "../../../const/ConfigNameConst";
import { GrainCarConst } from "../../../const/GrainCarConst";
import { TimeConst } from "../../../const/TimeConst";
import { G_ConfigLoader, G_ServerTime } from "../../../init";
import { assert } from "../../../utils/GlobleFunc";



export default class GrainCarConfigHelper {

  
   static  getGrainCarConfig(id) {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR).get(id);
        assert(info, 'graincar config can not find id = %d');
        return info;
    };
   static  getGrainCarWeek() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_WEEK);
        assert(info, 'graincar parameter config can not find id = %d');
        var array = info.content.split('|');
        var result = [];
        for (let i in array) {
            var v = array[i];
            result.push(parseInt(v));
        }
        return result;
    };
   static  getGrainCarOpenTime() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_OPEN);
        assert(info, 'graincar parameter config can not find id = %d');
        var array = info.content.split('|');
        return [
            parseInt(array[0]),
            parseInt(array[1])
        ];
    };
   static  getGrainCarCloseTime() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_CLOSE);
        assert(info, 'graincar parameter config can not find id = %d');
        var array = info.content.split('|');
        return [
            parseInt(array[0]),
            parseInt(array[1])
        ];
    };
   static  getGrainCarRouteGenTime() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_ROUTE);
        assert(info, 'graincar parameter config can not find id = %d');
        var array = info.content.split('|');
        return [
            parseInt(array[0]),
            parseInt(array[1])
        ];
    };
   static  getGrainCarStopTime(mineType) {
        var id = GrainCarConst.GRAINCAR_COPPER_STOP;
        if (mineType == GrainCarConst.MINE_COLOR_COPPER) {
            id = GrainCarConst.GRAINCAR_COPPER_STOP;
        } else if (mineType == GrainCarConst.MINE_COLOR_SILVER) {
            id = GrainCarConst.GRAINCAR_SILVER_STOP;
        } else if (mineType == GrainCarConst.MINE_COLOR_GOLD) {
            id = GrainCarConst.GRAINCAR_GOLD_STOP;
        }
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(id);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarMoveTime() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_MOVING_TIME);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarDonateCost() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_DONATE_COST);
        assert(info, 'graincar parameter config can not find id = %d');
        var array = info.content.split('|');
        return [
            parseInt(array[0]),
            parseInt(array[1]),
            parseInt(array[2])
        ];
    };
   static  getGrainCarDonateExp() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_DONATE_EXP);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarAttackTimes() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_ATTACK_TIMES);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarAttackBonus() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_ATTACK_BONUS);
        assert(info, 'graincar parameter config can not find id = %d');
        var array = info.content.split('|');
        return [
            parseInt(array[0]),
            parseInt(array[1]),
            parseInt(array[2])
        ];
    };
   static  getGrainCarAttackCD() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_ATTACK_CD);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarAttackHurt() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_ATTACK_HURT);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarDamageBonus() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_DAMAGE_BONUS);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarDamageReduce(level) {
        var id = GrainCarConst['GRAINCAR_PROTECT_' + level];
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(id);
        assert(info, 'graincar parameter config can not find id = %d');
        var array = info.content.split('|');
        return [
            parseInt(array[0]),
            parseInt(array[1])
        ];
    };
   static  getGrainCarMaxPlayerNum() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_MAX_NUM);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarShowTimeAfterEnd() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_SHOW);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarAttackLose() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_LOSE);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarShowHero() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_SHOW_HERO);
        assert(info, 'graincar parameter config can not find id = %d');
        var array = info.content.split('|');
        return array;
    };
   static  getGrainCarShame() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_SHOW_SHAME);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarLevelUp() {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_PARAMETER).get(GrainCarConst.GRAINCAR_LEVEL_UP);
        assert(info, 'graincar parameter config can not find id = %d');
        return parseInt(info.content);
    };
   static  getGrainCarRouteWithId(id) {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_ROUTE).get(id);
        assert(info, 'graincar route config can not find id = %d');
        var index = 1;
        var route = [];
        while (index <= GrainCarConst.ROUTE_STOP_COUNT_MAX && info['point_' + index] != 0) {
            route.push(info['point_' + index]);
            index = index + 1;
        }
        return [
            info.start,
            route[route.length],
            route
        ];
    };
   static  getGrainCarOpenTimeSecond() {
        var [hour, minute] = GrainCarConfigHelper.getGrainCarOpenTime();
        return hour * 3600 + minute * 60;
    };
   static  getGrainCarOpenTimeStamp() {
        var [hour, minute] = GrainCarConfigHelper.getGrainCarOpenTime();
        return G_ServerTime.getTimestampByHMS(hour, minute);
    };
   static  getGrainCarEndTimeStamp() {
        var [hour, minute] = GrainCarConfigHelper.getGrainCarCloseTime();
        return G_ServerTime.getTimestampByHMS(hour, minute);
    };
   static  getGrainCarGenTimeStamp() {
        var [genStartHour, genStartMinute] = GrainCarConfigHelper.getGrainCarRouteGenTime();
        return G_ServerTime.getTimestampByHMS(genStartHour, genStartMinute);
    };
   static  getNextGrainCarStartTime() {
        var weekArray = GrainCarConfigHelper.getGrainCarWeek();
        var weekRefreshTime = [];
        var grainCarOpenTimeSecond = GrainCarConfigHelper.getGrainCarOpenTimeSecond();
        for (let k in weekArray) {
            var v = weekArray[k];
            var time = G_ServerTime.getTimeByWdayandSecond(v + 1, grainCarOpenTimeSecond);
            weekRefreshTime.push(time);
        }
        var nextRefreshTime = 0;
        var index = 1;
        var curTime = G_ServerTime.getTime();
        while (index <= weekRefreshTime.length) {
            if (curTime < weekRefreshTime[index]) {
                nextRefreshTime = weekRefreshTime[index];
                break;
            }
            index = index + 1;
        }
        if (index == weekRefreshTime.length + 1) {
            nextRefreshTime = weekRefreshTime[1] + TimeConst.SECONDS_ONE_WEEK;
        }
        return nextRefreshTime;
    };
   static  isInActivityTime() {
        var curSecond = G_ServerTime.secondsFromToday();
        var [startHour, startMinute] = GrainCarConfigHelper.getGrainCarOpenTime();
        var startSecond = startHour * 3600 + startMinute * 60;
        var [endHour, endMinute] = GrainCarConfigHelper.getGrainCarCloseTime();
        var endSecond = endHour * 3600 + endMinute * 60 + GrainCarConfigHelper.getGrainCarShowTimeAfterEnd();
        return curSecond >= startSecond && curSecond < endSecond;
    };
   static  isInLaunchTime() {
        var curSecond = G_ServerTime.secondsFromToday();
        var [startHour, startMinute] = GrainCarConfigHelper.getGrainCarOpenTime();
        var startSecond = startHour * 3600 + startMinute * 60;
        var [endHour, endMinute] = GrainCarConfigHelper.getGrainCarCloseTime();
        var endSecond = endHour * 3600 + endMinute * 60;
        return curSecond >= startSecond && curSecond < endSecond;
    };
   static  isInActivityTimeFromGenerate() {
        if (!GrainCarConfigHelper.isTodayOpen()) {
            return false;
        }
        var curSecond = G_ServerTime.secondsFromToday();
        var [genStartHour, genStartMinute] = GrainCarConfigHelper.getGrainCarRouteGenTime();
        var genStartSecond = genStartHour * 3600 + genStartMinute * 60;
        var [endHour, endMinute] = GrainCarConfigHelper.getGrainCarCloseTime();
        var endSecond = endHour * 3600 + endMinute * 60 + GrainCarConfigHelper.getGrainCarShowTimeAfterEnd();
        return curSecond >= genStartSecond && curSecond < endSecond;
    };
   static  getOpenDays() {
        var weekArray = GrainCarConfigHelper.getGrainCarWeek();
        var days = {};
        for (let k in weekArray) {
            var v = weekArray[k];
            var curDay = parseInt(v);
            if (curDay > 0) {
                if (curDay == 7) {
                    curDay = 0;
                }
                days[curDay] = true;
            }
        }
        return days;
    };
   static  isTodayOpen() {
        var ret = false;
        var date = G_ServerTime.getDateObject(null, 0);
        var days = GrainCarConfigHelper.getOpenDays();
        var curDay = date.getDay();
        if (days[curDay]) {
            return !ret;
        }
        return ret;
    };
   static  isClose() {
        var curSecond = G_ServerTime.secondsFromToday();
        var [endHour, endMinute] = GrainCarConfigHelper.getGrainCarCloseTime();
        var closeSecond = endHour * 3600 + endMinute * 60;
        var endSecond = endHour * 3600 + endMinute * 60 + GrainCarConfigHelper.getGrainCarShowTimeAfterEnd();
        return curSecond >= closeSecond && curSecond < endSecond;
    };
   static  getActStage() {
        var startTime = GrainCarConfigHelper.getGrainCarOpenTimeStamp().getTime()/1000;
        var endTime = GrainCarConfigHelper.getGrainCarEndTimeStamp().getTime()/1000;
        var genTime = GrainCarConfigHelper.getGrainCarGenTimeStamp().getTime()/1000;
        var curTime = G_ServerTime.getTime();
        if (curTime >= genTime && curTime < startTime) {
            return GrainCarConst.ACT_STATGE_GENERATED;
        } else if (curTime >= startTime && curTime < endTime) {
            return GrainCarConst.ACT_STATGE_OPEN;
        } else {
            return GrainCarConst.ACT_STATGE_CLOSE;
        }
    };
   static  getRouteWithId(id) {
        var info = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR_ROUTE).get(id);
        assert(info, 'graincar_route config can not find id = %d');
        var route = [];
        route.push(info.start);
        var i = 1;
        while (i < 13 && info['point_' + i] > 0) {
            route.push(info['point_' + i]);
            i = i + 1;
        }
        return route;
    };
   static  randDiff(startNum, endNum, count) {
        var fillNum = function (m, n) {
            var j, k;
            if (n) {
                j = m;
                k = n;
            } else {
                j = 1;
                k = m;
            }
            var t = [];
            for (var i = j; i <= k; i++) {
                t.push(i);
            }
            return t;
        }
        var tmp = fillNum(startNum, endNum);
        if (count > endNum - startNum + 1) {
            return tmp;
        }
        var x = 0;
        var t = [];
        var getRandom = function(end){
            return Math.floor(Math.random()*end)
        }
        while (count > 0) {
            x = getRandom(endNum - startNum + 1);
            if(x>=tmp.length)x=tmp.length-1;
            if(x<0)x = 0;
            t.push(tmp[x]);
            tmp.splice(x, 1);
            count = count - 1;
            startNum = startNum + 1;
        }
        return t;
    };
}

File: study/lua/src/app/scene/view/grainCar/GrainCarConfigHelper.lua
local GrainCarConfigHelper = {}

local GrainCarParameter = require("app.config.graincar_parameter")
local GrainCarRoute = require("app.config.graincar_route")
local GrainCarConst  = require("app.const.GrainCarConst")
local TimeConst  = require("app.const.TimeConst")

------------------------------------------------------------
--------------graincar 配置-----------------------
------------------------------------------------------------
function GrainCarConfigHelper.getGrainCarConfig(id)
    local info = require("app.config.graincar").get(id)
	assert(info, string.format("graincar config can not find id = %d", id))
	return info
end


------------------------------------------------------------
--------------graincar_parameter 配置-----------------------
------------------------------------------------------------
--每周几开启 
--return: [1, 3, 5]
function GrainCarConfigHelper.getGrainCarWeek()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_WEEK)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_WEEK))
    local array = string.split(info.content, "|")
    local result = {}
    for i, v in ipairs(array) do
        table.insert(result, tonumber(v))
    end
	return result
end

--活动开始时间 
--return: hour, minute
function GrainCarConfigHelper.getGrainCarOpenTime()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_OPEN)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_OPEN))
    local array = string.split(info.content, "|")
	return tonumber(array[1]), tonumber(array[2])
end

--活动结束时间 
--return: hour, minute
function GrainCarConfigHelper.getGrainCarCloseTime()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_CLOSE)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_CLOSE))
    local array = string.split(info.content, "|")
	return tonumber(array[1]), tonumber(array[2])
end

--粮车路线生成时间 
--return: hour, minute
function GrainCarConfigHelper.getGrainCarRouteGenTime()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_ROUTE)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_ROUTE))
    local array = string.split(info.content, "|")
	return tonumber(array[1]), tonumber(array[2])
end

--粮车停留时间
--param mineType: GrainCarConst.MINE_COLOR_COPPER 铜矿
--                GrainCarConst.MINE_COLOR_SILVER 银矿
--                GrainCarConst.MINE_COLOR_GOLD   金矿
--return: second
function GrainCarConfigHelper.getGrainCarStopTime(mineType)
    local id = GrainCarConst.GRAINCAR_COPPER_STOP
    if mineType == GrainCarConst.MINE_COLOR_COPPER then
        id = GrainCarConst.GRAINCAR_COPPER_STOP
    elseif mineType == GrainCarConst.MINE_COLOR_SILVER then
        id = GrainCarConst.GRAINCAR_SILVER_STOP
    elseif mineType == GrainCarConst.MINE_COLOR_GOLD then
            id = GrainCarConst.GRAINCAR_GOLD_STOP
    end
    local info = GrainCarParameter.get(id)
    assert(info, string.format("graincar parameter config can not find id = %d", id))
	return tonumber(info.content)
end

--粮车移动时间（2个矿间）
--return: second
function GrainCarConfigHelper.getGrainCarMoveTime()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_MOVING_TIME)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_MOVING_TIME))
	return tonumber(info.content)
end

--粮车捐献消耗 军团贡献
--return: type, value, size
function GrainCarConfigHelper.getGrainCarDonateCost()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_DONATE_COST)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_DONATE_COST))
    local array = string.split(info.content, "|")
	return tonumber(array[1]), tonumber(array[2]), tonumber(array[3])
end

--捐献获得升级经验
--return: int
function GrainCarConfigHelper.getGrainCarDonateExp()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_DONATE_EXP)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_DONATE_EXP))
	return tonumber(info.content)
end

--获得奖励，需要攻击粮车几次
--return: int
function GrainCarConfigHelper.getGrainCarAttackTimes()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_ATTACK_TIMES)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_ATTACK_TIMES))
	return tonumber(info.content)
end

--攻击粮车获得奖励 军团贡献
--return: type, value, size
function GrainCarConfigHelper.getGrainCarAttackBonus()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_ATTACK_BONUS)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_ATTACK_BONUS))
    local array = string.split(info.content, "|")
	return tonumber(array[1]), tonumber(array[2]), tonumber(array[3])
end

--攻击粮车间隔 秒
--return: second
function GrainCarConfigHelper.getGrainCarAttackCD()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_ATTACK_CD)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_ATTACK_CD))
	return tonumber(info.content)
end

--每次攻击粮车损失耐久度
--return: int
function GrainCarConfigHelper.getGrainCarAttackHurt()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_ATTACK_HURT)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_ATTACK_HURT))
	return tonumber(info.content)
end

--粮车损毁获得奖励百分比
--return: int
function GrainCarConfigHelper.getGrainCarDamageBonus()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_DAMAGE_BONUS)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_DAMAGE_BONUS))
	return tonumber(info.content)
end

--粮车伤害减免
--当大于等于x名本军团成员和粮车在1个矿点中时，矿车受攻击时损失的耐久值减少x%
--param level:1,2,3 档
--return: int
function GrainCarConfigHelper.getGrainCarDamageReduce(level)
    local id = GrainCarConst["GRAINCAR_PROTECT_" .. level]
    local info = GrainCarParameter.get(id)
    assert(info, string.format("graincar parameter config can not find id = %d", id))
    local array = string.split(info.content, "|")
	return tonumber(array[1]), tonumber(array[2])
end

--同屏显示最大人数
function GrainCarConfigHelper.getGrainCarMaxPlayerNum()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_MAX_NUM)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_MAX_NUM))
	return tonumber(info.content)
end

--结束后还显示多久
function GrainCarConfigHelper.getGrainCarShowTimeAfterEnd()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_SHOW)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_SHOW))
	return tonumber(info.content)
end

--攻击粮车后损失多少兵力
function GrainCarConfigHelper.getGrainCarAttackLose()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_LOSE)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_LOSE))
	return tonumber(info.content)
end

--护镖的avatar
function GrainCarConfigHelper.getGrainCarShowHero()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_SHOW_HERO)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_LOSE))
    local array = string.split(info.content, "|")
    return array
end

--活动结束后，显示死亡粮车的时间
function GrainCarConfigHelper.getGrainCarShame()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_SHOW_SHAME)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_SHOW_SHAME))
    return tonumber(info.content)
end

--捐献几人次后才能发车
function GrainCarConfigHelper.getGrainCarLevelUp()
    local info = GrainCarParameter.get(GrainCarConst.GRAINCAR_LEVEL_UP)
    assert(info, string.format("graincar parameter config can not find id = %d", GrainCarConst.GRAINCAR_LEVEL_UP))
    return tonumber(info.content)
end

------------------------------------------------------------
------------------graincar_route 配置-----------------------
------------------------------------------------------------
--获取起点，终点，路线
--return 起点， 终点， 路线列表[108, 109, ... , 201]
function GrainCarConfigHelper.getGrainCarRouteWithId(id)
    local info = GrainCarRoute.get(id)
    assert(info, string.format("graincar route config can not find id = %d", id))
    local index = 1
    local route = {}
    while index <= GrainCarConst.ROUTE_STOP_COUNT_MAX and info["point_" .. index] ~= 0 do
        table.insert(route, info["point_" .. index])
        index = index + 1
    end
    
	return info.start, route[#route], route
end



------------------------------------------------------------
--------------------------方法-------------------------------
------------------------------------------------------------
--活动开始时间 相对0点偏移多少秒
function GrainCarConfigHelper.getGrainCarOpenTimeSecond()
    local hour, minute = GrainCarConfigHelper.getGrainCarOpenTime()
    return hour * 3600 + minute * 60
end

--获取活动开始时间戳
function GrainCarConfigHelper.getGrainCarOpenTimeStamp()
    local hour, minute = GrainCarConfigHelper.getGrainCarOpenTime()
    return G_ServerTime:getTimestampByHMS(hour, minute)
end

--获取活动结束时间戳
function GrainCarConfigHelper.getGrainCarEndTimeStamp()
    local hour, minute = GrainCarConfigHelper.getGrainCarCloseTime()
    return G_ServerTime:getTimestampByHMS(hour, minute)
end

--获取粮车生成时间戳
function GrainCarConfigHelper.getGrainCarGenTimeStamp()
    local genStartHour, genStartMinute = GrainCarConfigHelper.getGrainCarRouteGenTime()
    return G_ServerTime:getTimestampByHMS(genStartHour, genStartMinute)
end

--获取下次活动开始时间戳
function GrainCarConfigHelper.getNextGrainCarStartTime()
    local weekArray = GrainCarConfigHelper.getGrainCarWeek()

    local weekRefreshTime = {}
    local grainCarOpenTimeSecond = GrainCarConfigHelper.getGrainCarOpenTimeSecond()
    for k, v in pairs(weekArray) do
        local time = G_ServerTime:getTimeByWdayandSecond(v + 1, grainCarOpenTimeSecond)
        table.insert(weekRefreshTime, time)
    end

    local nextRefreshTime = 0
    local index = 1
    local curTime = G_ServerTime:getTime()
    while index <= #weekRefreshTime do
        if curTime < weekRefreshTime[index] then
            nextRefreshTime =  weekRefreshTime[index]
            break
        end
        index = index + 1
    end
    if index == #weekRefreshTime + 1 then
        nextRefreshTime = weekRefreshTime[1] + TimeConst.SECONDS_ONE_WEEK
    end

    return nextRefreshTime
end

-- 是否在活动时间
function GrainCarConfigHelper.isInActivityTime()
    local curSecond = G_ServerTime:secondsFromToday()
    local startHour, startMinute = GrainCarConfigHelper.getGrainCarOpenTime()
    local startSecond = startHour * 3600 + startMinute * 60
    local endHour, endMinute = GrainCarConfigHelper.getGrainCarCloseTime()
    local endSecond = endHour * 3600 + endMinute * 60 + GrainCarConfigHelper.getGrainCarShowTimeAfterEnd()
    return (curSecond >= startSecond and curSecond < endSecond)
end

-- 是否在发车时间区间
function GrainCarConfigHelper.isInLaunchTime()
    local curSecond = G_ServerTime:secondsFromToday()
    local startHour, startMinute = GrainCarConfigHelper.getGrainCarOpenTime()
    local startSecond = startHour * 3600 + startMinute * 60
    local endHour, endMinute = GrainCarConfigHelper.getGrainCarCloseTime()
    local endSecond = endHour * 3600 + endMinute * 60
    return (curSecond >= startSecond and curSecond < endSecond)
end

-- 是否在活动时间(从粮车生成开始算)
function GrainCarConfigHelper.isInActivityTimeFromGenerate()
    if not GrainCarConfigHelper.isTodayOpen() then
        return false
    end
    local curSecond = G_ServerTime:secondsFromToday()
    local genStartHour, genStartMinute = GrainCarConfigHelper.getGrainCarRouteGenTime()
    local genStartSecond = genStartHour * 3600 + genStartMinute * 60
    local endHour, endMinute = GrainCarConfigHelper.getGrainCarCloseTime()
    local endSecond = endHour * 3600 + endMinute * 60 + GrainCarConfigHelper.getGrainCarShowTimeAfterEnd()
    return (curSecond >= genStartSecond and curSecond < endSecond)
end

function GrainCarConfigHelper.getOpenDays()
    local weekArray = GrainCarConfigHelper.getGrainCarWeek()
    local days = {}
    for k, v in pairs(weekArray) do
        local curDay = tonumber(v)
        if curDay > 0 then
            curDay = curDay + 1
            if curDay > 7 then
                curDay = 1
            end
            days[curDay] = true
        end
    end

    return days
end

-- 是否今日开启
function GrainCarConfigHelper.isTodayOpen()
    local ret = false
    local date = G_ServerTime:getDateObject(nil, 0)
    local days = GrainCarConfigHelper.getOpenDays()
    if days[date.wday] then
        return not ret
    end
    return ret
end

--是否活动结束
function GrainCarConfigHelper.isClose()
    local curSecond = G_ServerTime:secondsFromToday()
    local endHour, endMinute = GrainCarConfigHelper.getGrainCarCloseTime()
    local closeSecond = endHour * 3600 + endMinute * 60
    local endSecond = endHour * 3600 + endMinute * 60 + GrainCarConfigHelper.getGrainCarShowTimeAfterEnd()
    return (curSecond >= closeSecond and curSecond < endSecond)
end

--活动阶段
function GrainCarConfigHelper.getActStage()
    local startTime = GrainCarConfigHelper.getGrainCarOpenTimeStamp()
    local endTime = GrainCarConfigHelper.getGrainCarEndTimeStamp()
    local genTime = GrainCarConfigHelper.getGrainCarGenTimeStamp()
    local curTime = G_ServerTime:getTime()
    if curTime >= genTime and curTime < startTime then
		return GrainCarConst.ACT_STATGE_GENERATED
	elseif curTime >= startTime and curTime < endTime then
		return GrainCarConst.ACT_STATGE_OPEN
	else
		return GrainCarConst.ACT_STATGE_CLOSE
	end
end

--获取经过路径
function GrainCarConfigHelper.getRouteWithId(id)
    local info = require("app.config.graincar_route").get(id)
    assert(info, string.format("graincar_route config can not find id = %d", id))
    local route = {}
    table.insert(route, info.start)
    local i = 1
    while i < 13 and info["point_" .. i] > 0 do
        table.insert(route, info["point_" .. i])
        i = i + 1
    end
    return route
end

--在startNum~endNum范围内随机出count个数字
function GrainCarConfigHelper.randDiff(startNum, endNum, count)
    --产生1~~m,若有n的则m~~n的数字表
    local function fillNum(m, n)
        local j,k
        if n then
            j = m
            k = n
        else
            j=1
            k=m
        end
        local t={}
        for i=j, k do
            table.insert(t, i)
        end
        return t
    end

    local tmp = fillNum(startNum, endNum)
    if count > endNum - startNum + 1 then
        return tmp
    end
    local x = 0
    local t = {}
	
    while count > 0 do
       x = math.random(1, endNum - startNum + 1)
       table.insert(t, tmp[x])
       table.remove(tmp, x)
       count = count-1
       startNum = startNum + 1
    end
    return t
end

return GrainCarConfigHelper